### DISCUSSION TITLE
Triangle context

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#92] [Q&A] [answered]
I create a new triangle context for debug drawing shapes and keep it alive until the drawing is taking place:

```js
const triContext = new Jolt.ShapeGetTriangles(shape, Jolt.AABox.prototype.sBiggest(), shape.GetCenterOfMass(), Jolt.Quat.prototype.sIdentity(), scale);
```

This works fine for general shapes. However, not for soft bodies, where the vertices change after every update. In case of soft bodies, the context is also needed to get the vertex positions in order to be able to update the visual mesh. Which brings me to question:

When creating a `triContext` for a soft body, I couldn't find a way to re-use one, so I destroy the old one and create a new one every update. Is it correct?

**Comment by jrouwe** [comment, marked as answer]
> When creating a triContext for a soft body, I couldn't find a way to re-use one, so I destroy the old one and create a new one every update. Is it correct?

Yes it is, and this is how the [demo](https://github.com/jrouwe/JoltPhysics.js/blob/ddda56a159504f7708045bf3c094fc9e771553a7/Examples/js/example.js#L164-L165) is doing it as [well](https://github.com/jrouwe/JoltPhysics.js/blob/ddda56a159504f7708045bf3c094fc9e771553a7/Examples/js/example.js#L231-L250). Don't forget to call ```Jolt.destroy(triContext)``` though or else you'll run out of memory very quickly.

B.t.w. this is not the most efficient way of doing things. You have direct access to the vertex data using something like:

```
Jolt.castObject(body.GetMotionProperties(), Jolt.SoftBodyMotionProperties).GetVertices()
```

(except that I see now that I didn't expose SoftBodyMotionProperties to JS yet, I'll do that tomorrow)

**Reply by jrouwe** [reply]
Done. I also added a checkbox to the [sample](https://jrouwe.github.io/JoltPhysics.js/soft_body.html) to show it off.

**Reply by LeXXik** [reply]
Thank you!

**Comment by LeXXik** [comment]
I've now implemented the approach with casting to `SoftBodyMotionProperties`. However, I am curious about your earlier comment:

> this is not the most efficient way of doing things

Could you please elaborate a little? From what it looks like casting to `SoftBodyMotionProperties` loops through the vertices, while the previous method I think did not? Maybe I missed something.

**Reply by jrouwe** [reply]
```ShapeGetTriangles``` internally does the same loop and copies it to a buffer, you then need to copy it to your own buffer so effectively you're allocating 2 large buffers and copying the data 2x. It's hard to say which of the methods is faster since a call from JS to WASM also brings some overhead, so accessing a vertex incurs some penalty too. In any case I think that most of the time you don't just want to render the physics data but you probably want to render a textured mesh so it is nice to have access to the vertex positions in the order they were provided so you know what's what.

**Reply by LeXXik** [reply]
Yep, works good. The mapping is straight forward as well. Here is my attempt (2 substeps, 1/30 timestep):
https://playcanv.as/b/e8bd5bc9

**Comment by LeXXik** [comment]
I noticed the objects fall through the faces in a soft body, if the triangle is larger than the object. Is this expected? Here, I have only 4 vertices for the soft body (3 corners are fixed, 1 free):

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/b9bbf803-32a4-4ebd-afb7-b93cb76dde76

For this sample, I am hardcoding the face indices:

```js
settings.AddFace(new Jolt.SoftBodySharedSettingsFace(2, 1, 0, 0));
settings.AddFace(new Jolt.SoftBodySharedSettingsFace(2, 3, 1, 0));
```

**Reply by jrouwe** [reply]
Yes, only the particles of the cloth do collision while simulating
